paderbox/io/download.py

```diff
@@ -0,0 +1,184 @@
+
+import os
+import socket
+import sys
+import tarfile
+import zipfile
+from os import path
+from urllib.request import urlretrieve
+
+from tqdm import tqdm
+
+
+def download_file(remote_file, local_file):
+    """
+    Download single file to local_dir
+
+    Args:
+        remote_file:
+        local_file:
+
+    Returns:
+
+    """
+    local_dir = path.dirname(local_file)
+    if not path.isfile(local_file):
+        def progress_hook(t):
+            """
+            Wraps tqdm instance. Don't forget to close() or __exit__()
+            the tqdm instance once you're done with it (easiest using
+            `with` syntax).
+            """
+
+            last_b = [0]
+
+            def inner(b=1, bsize=1, tsize=None):
+                """
+                b  : int, optional
+                    Number of blocks just transferred [default: 1].
+                bsize  : int, optional
+                    Size of each block (in tqdm units) [default: 1].
+                tsize  : int, optional
+                    Total size (in tqdm units). If [default: None]
+                    remains unchanged.
+                """
+                if tsize is not None:
+                    t.total = tsize
+                t.update((b - last_b[0]) * bsize)
+                last_b[0] = b
+
+            return inner
+
+        tmp_file = path.join(local_dir, 'tmp_file')
+        with tqdm(
+                desc="{0: >25s}".format(
+                    path.splitext(remote_file.split('/')[-1])[0]),
+                file=sys.stdout,
+                unit='B',
+                unit_scale=True,
+                miniters=1,
+                leave=False,
+                ascii=True
+        ) as t:
+
+            urlretrieve(
+                remote_file,
+                filename=tmp_file,
+                reporthook=progress_hook(t),
+                data=None
+            )
+        os.rename(tmp_file, local_file)
+    return local_file
+
+
+def extract_file(local_file):
+    """
+    extract files if local_file zip or tar.gz else do nothing
```

Comment:
Is my understanding correct, that this code skips existing files?

Could you add a description to the doctext for the `# Trick to omit first level folder`?

Reply:
Could you also add a motivation for the `# Trick to omit first level folder`?